### PR TITLE
Added the ability to set attributes on a HalLink

### PR DESCRIPTION
**Pull request by iainmckay**
We had a use case where we needed to be able to set attributes on a link that was already added to a collection. This PR adds the ability to set attributes on a HalLink.

### CODE REVIEW

**Comment by blongden** [comment]
I'm so conflicted with this one. One one hand I think, "meh, it's what people want". On the other I think, "Meep, mutable state!!".

Not that state is immutable all over this lib, of course...

File: src/Nocarrier/HalLink.php [hunk 1]
```diff
@@ -80,6 +80,31 @@ public function getAttributes()
...
     /**
+     * Set a particular attribute on this link.
+     *
+     * @param string $name
+     *   Name of the attribute to set.
+     * @param mixed $value
+     *   Value to set the attribute to.
+     * @return void
+     */
+    public function setAttribute($name, $value)
```
Comment (hjr3) [inline src/Nocarrier/HalLink.php]: If we are going down this route, then I think this should have an `overwrite` flag. Mutating state after the fact should require people to be very explicit about what they are doing.

**Comment by weierophinney** [comment]
@blongden I'd argue that it breaks the existing semantics of the class, as it was immutable before, and the URI remains immutable after the change; it should either all be immutable, or all mutable.

@iainmckay Write a utility function that creates a new instance:

```php
function cloneWithAttributes(HalLink $link, array $attributes)
{
    return new HalLink($original->getUri(), $attributes);
}
```

And then pass the new link to the container under the same relation.

File: src/Nocarrier/HalLink.php [hunk 2]
```diff
@@ -80,6 +80,31 @@ public function getAttributes()
...
+        $this->attributes[$name] = $value;
+    }
+
+    /**
+     * Set the attributes on this link.
+     *
+     * @param array $attributes
+     *   Attibutes to set on this link.
+     */
+    public function setAttributes(array $attributes)
```
Comment (hjr3) [inline src/Nocarrier/HalLink.php]: Similar to my note above, I think this is asking for problems. I would make people use the above `setAttribute` method and be explicit about whether they are intentionally overwriting something.

**Comment by hjr3** [comment]
We can go down the `DateTime` path of creating a `MutableHalLink` class that extends `HalLink` too.

**Comment by iainmckay** [comment]
Yes, at the time of writing this I didn't understand that it was intended to be immutable. In our particular use case we inherit from `Hal` and attach extra pieces of data.

Looking at our code I can easily refactor out the need for this PR. Based on the discussion here it appears it would confuse things to merge it.

If no one has a problem, I'll close this PR.

**Comment by blongden** [comment]
@weierophinney That's a very good point about the URI being immutable

@iainmckay Thanks for the contribution anyway :dancer: :+1:

**Pull request closed without merging**